### PR TITLE
Feature/apl 652 add zip shard step

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
Added zipping step and updated DownloadableFilesManager + unit tests.
The 'merkle tree hash' is stored into SHARD record on early step.

### CODE REVIEW

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import org.jboss.weld.junit5.WeldInitiator;
+import org.jboss.weld.junit5.WeldSetup;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+
+@EnableWeld
+class DownloadableFilesManagerTest {
+    private static final Logger log = getLogger(DownloadableFilesManagerTest.class);
+    private Path csvResourcesPath = new ResourceFileLoader().getResourcePath().toAbsolutePath();
+    private final Bean<Path> dataExportDir = MockBean.of(csvResourcesPath, Path.class);
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManagerTest.java]: Unused bean

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/Zip.java [hunk 2]
```diff
@@ -1,150 +1,27 @@
...
+    /**
+     * Compress all filtered files in directory into ZIP file, change file timestamp to be predefined.
+     * Return computed CRC/hash for created ZIP as byte array.
+     *
+     * @param zipFile path to zip file
+     * @param inputFolder folder to zip
+     * @param filesTimeFromEpoch NULL (default will be assigned) or time in ms from Epoch for all file times
+     * @param filenameFilter NULL (CSV will be by default) or file filter instance
+     * @return byte array filled by CRC/hash if success, ZERO length array if something went wrong
+     * @throws RuntimeException if there are no CSF file inside specified folder
```
Comment (AndrewBoyarsky) [inline apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/Zip.java]: Fix Typo CSF->CSV

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+ * Class is used for zip-unzip filtered files in specified directory
+ *
+ * @author alukin@gmail.com
+ */
+@Singleton
+public class ZipImpl implements Zip {
+    private static final Logger log = LoggerFactory.getLogger(ZipImpl.class);
+
+    public final static int FILE_CHUNK_SIZE = 32768; // magic constant copied from DownloadableFilesManager class
+    private final static int BUF_SIZE= 1042 * 2; // 2 Mb
```
Comment (AndrewBoyarsky) [inline apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java]: 1024 * 2 = 2Kb vs 1042 * 2 = 2Mb

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public byte[] compress(String zipFile, String inputFolder, Long filesTimeFromEpoch,
+                            FilenameFilter filenameFilter) {
+        Objects.requireNonNull(zipFile, "zipFile is NULL");
+        Objects.requireNonNull(inputFolder, "inputFolder is NULL");
+        if (zipFile.isEmpty()) {
```
Comment (AndrewBoyarsky) [inline apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java]: Better to use StringValidator.requireNonBlank

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        long start = System.currentTimeMillis();
+        log.trace("Creating file '{}' in folder '{}', filesTimestamp = {}", zipFile, inputFolder, filesTimeFromEpoch);
+        byte[] zipCrcHash;
+        File directory = new File(inputFolder);
+        if (filenameFilter == null) {
+            filenameFilter = DEFAULT_CSV_FILE_FILTER;
+        }
+        // get file(s) listing from folder by filter (no recursion for subfolders(s) !)
+        File[] fileList = directory.listFiles(filenameFilter);
```
Comment (AndrewBoyarsky) [inline apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java]: General zip implementation should also zip subfolders

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java [hunk 6]
```diff
@@ -28,53 +34,61 @@
...
-    private String fileBaseDir="/home/at/testfiles";
+    private String fileBaseDir /*="/home/at/testfiles"*/;
     
     @Inject
     public DownloadableFilesManager(DirProvider dirProvider) {
-        fileBaseDir=dirProvider.getDbDir()+File.separator+FILES_SUBDIR;
+        Objects.requireNonNull(dirProvider, "dirProvider is NULL");
+        Objects.requireNonNull(dirProvider.getDataExportDir(), "dataExportDir in dirProvider is NULL");
+//        fileBaseDir=dirProvider.getDbDir()+File.separator+FILES_SUBDIR;
+        this.fileBaseDir = dirProvider.getDataExportDir().toString();
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java]: That was agreement.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java [hunk 7]
```diff
@@ -85,23 +99,49 @@ private FileDownloadInfo createFileDownloadInfo(String fileId) {
                     fci.present=3;
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java]: @oleksiylukin Can you say something ?

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java (hunk 5 above)
Comment (yuriy-larin-firstbridge) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java]: No need to zip sub-folder now, may be later when we have it.

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java (hunk 4 above)
Comment (yuriy-larin-firstbridge) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java]: Done

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/ZipImpl.java]: set up 16 Kb

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/Zip.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/Zip.java]: Done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManagerTest.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManagerTest.java]: Removed

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/commands/CommandParamInfo.java [hunk 8]
```diff
@@ -34,4 +34,10 @@
 
     void setDbIdExclusionSet(Set<Long> dbIdExclusionSet);
 
+    /**
+     * To check if getShardHash() returns 'merkle tree hash' or 'zip crc hash'
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/commands/CommandParamInfo.java]: Field value is needed at different time. If we will need to have them both at the same time, we should create two separate fields. Now it's ok.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java [hunk 9]
```diff
@@ -521,12 +529,53 @@ public MigrateState archiveCsv(CommandParamInfo paramInfo) {
...
+                    dirProvider.getDataExportDir().toAbsolutePath().toString(), null, null);
+
+            // prepare real CRC data for shard record update
+            paramInfo = new CommandParamInfoImpl(zipCrcHash, true);
+            updateShardRecord(paramInfo, sourceDataSource, state, 1L); //update shard record by ZIP crc value
+
+            // update recovery
+            state = ZIP_ARCHIVE_FINISHED;
+            updateShardRecoveryProcessedTableList(sourceConnect, shardFileName, state);
+            sourceConnect.commit();
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java [hunk 10]
```diff
@@ -521,12 +529,53 @@ public MigrateState archiveCsv(CommandParamInfo paramInfo) {
             recovery = new ShardRecovery(ZIP_ARCHIVE_STARTED);
         }
 
-        // TODO: put into archive CSV for all : derived tables, shard, block_index, transaction_shard_index + compute CRC/hash and update column value in SHARD
-        // TODO: keep recovery data on export
-
-        state = ZIP_ARCHIVE_FINISHED;
-//        updateToFinalStepState(sourceDataSource.begin(), recovery, state);
-        log.debug("ZIP ARCHIVE Processed table(s)=[{}] in {} secs", paramInfo.getTableNameList().size(), (System.currentTimeMillis() - startAllTables)/1000);
+        if (createdShardId.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java]: Removed

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java [hunk 11]
```diff
@@ -85,23 +99,49 @@ private FileDownloadInfo createFileDownloadInfo(String fileId) {
...
+            String realShardId = fileId.substring(fileId.lastIndexOf("::") + 2);
+            long shardId = 0;
+            try {
+                shardId = Long.valueOf(realShardId);
+            } catch (NumberFormatException e) {
+                log.warn("Incorrect shardId value found in parameter = '{}'", fileId);
+                return null;
+            }
+            String fileName = ShardNameHelper.getShardArchiveNameByShardId(shardId);
+            absPath = this.fileBaseDir + File.separator + fileName + ".zip";
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java]: Path is resolved below by call
`Path res = Paths.get(absPath);`

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java]: Updated to ENUM

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java [hunk 12]
```diff
@@ -28,53 +34,61 @@
...
-            res.fileInfo.isPresent=true;
-            res.fileInfo.fileId=fileId;
-            res.created=new Date();
+        Objects.requireNonNull(fileId, "fileId is NULL");
+        FileDownloadInfo downloadInfo = new FileDownloadInfo();
+        Path fpath = mapFileIdToLocalPath(fileId);
+        if(fpath != null){
+            downloadInfo.fileInfo.isPresent = true;
+            downloadInfo.fileInfo.fileId = fileId;
+            downloadInfo.created = new Date();
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/DownloadableFilesManager.java]: UTC now

**Pull request merged**